### PR TITLE
[UNR-5651] 4.27 support

### PR DESCRIPTION
**Pull request by mironec**
#### Description
4.27 support.

#### Release note
Need to do.

#### Tests
A lot of tests, still need to do more.

#### Documentation
Still lot to do for our upgrade guide...

#### Primary reviewers
idk

Engine PR: https://github.com/improbableio/UnrealEngine/pull/1037
TestGym PR: https://github.com/spatialos/UnrealGDKTestGyms/pull/464
EP PR: https://github.com/spatialos/UnrealGDKExampleProject/pull/297

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp [hunk 1]
```diff
@@ -356,31 +356,26 @@ TSharedPtr<FUnrealType> CreateUnrealTypeInfo(UStruct* Type, uint32 ParentChecksu
 
 		TSharedPtr<FUnrealProperty> PropertyNode = nullptr;
 
+		// It's possible to have duplicate parent properties (they are distinguished by ArrayIndex), so we make sure to look at them
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp:359]: Why this change?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.h [hunk 2]
```diff
@@ -15,10 +15,10 @@ struct FControllerPawnPair
 	GENERATED_BODY()
 
 	UPROPERTY()
-	ATestPossessionPlayerController* PlayerController;
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.h:18]: Is this required now? A drive-by or..?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp (hunk 1 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp:359]: Was planning to discuss this somewhere, but may as well put some sort of explanation here.
This one is potentially dangerous, and I am not sure what I have done here will actually work. Previously to figure out if commands corresponded to funrealproperties, we used the CompatibleChecksum. Unfortunately, this is now computed differently for Cmds and Properties in 4.27, so this method will not work. Instead, I noticed they pointed to the same FProperty, but I don't know if that will always be the case, or if that even is a valid way of finding matching pairs. This needs quite a bit more investigation.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.h (hunk 2 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.h:18]: We now get spammed with a lot of "property is uninitialized" warnings/logs. There are actually about 20 more or something I still have to do. Would like to know what changed, would have to dig.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.h:18]: (Maybe nulling of UProps is handled differently now? Only for certain primitive types? No idea)

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.h:18]: Maybe a bit pedantic but these nullptr fixes could maybe go in separate PR as it shouldn't be controversial and can just go into master, whereas the other stuff needs debate as you say.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp (hunk 1 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp:359]: tricksy.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]: good news, found out I was dumb, and this is not needed

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]: Glad you got to the bottom of it!

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp [hunk 3]
```diff
@@ -79,6 +79,21 @@ uint32 GenerateChecksum(GDK_PROPERTY(Property) * Property, uint32 ParentChecksum
 	Checksum = FCrc::StrCrc32(*Property->GetCPPType(nullptr, 0).ToLower(), Checksum); // Evolve by property type
 	Checksum = FCrc::MemCrc32(&StaticArrayIndex, sizeof(StaticArrayIndex),
 							  Checksum); // Evolve by StaticArrayIndex (to make all unrolled static array elements unique)
+#if ENGINE_MINOR_VERSION >= 27 || ENGINE_MAJOR_VERSION == 5
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp:82]:
```suggestion
#if UE_VERSION_NEWER_THAN(4,27,0)
```
I found a nicer solution for this ^

Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp:82]:
```suggestion
#if UE_VERSION_NEWER_THAN(4,26,0)
```
Found a nicer way of doing this ^.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]: nice, unfortunate that it needs an extra include, they could have just put it next to the engine version definitions, I think, but I switched to using it. (Also needed 4.27.-1, which is a bit annoying, but ok)

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp [hunk 4]
```diff
@@ -79,6 +80,21 @@ uint32 GenerateChecksum(GDK_PROPERTY(Property) * Property, uint32 ParentChecksum
 	Checksum = FCrc::StrCrc32(*Property->GetCPPType(nullptr, 0).ToLower(), Checksum); // Evolve by property type
 	Checksum = FCrc::MemCrc32(&StaticArrayIndex, sizeof(StaticArrayIndex),
 							  Checksum); // Evolve by StaticArrayIndex (to make all unrolled static array elements unique)
+#if UE_VERSION_NEWER_THAN(4, 27, -1)
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp:83]: nice way of making an inclusive range!

File: CHANGELOG.md [hunk 5]
```diff
@@ -12,13 +12,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Breaking changes:
 - SpatialSwitchHasAuthority now respects World's version of IsServer which assumes server status when NetDriver is null.
+- We no longer support Unreal Engine version 4.25. We recommend that you upgrade to the newest version 4.27 to continue receiving updates.
 
 ### Features:
+- Unreal Engine version
```
Comment (joshuahuburn) [inline CHANGELOG.md:18]:
```suggestion
```

**Review by joshuahuburn — APPROVED** [review]
Cool job

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/CodeWriter.cpp [hunk 6]
```diff
@@ -38,7 +38,7 @@ FCodeWriter& FCodeWriter::Print(const FString& String)
 	// Replace 4 spaces with tabs.
 	for (auto& Line : Lines)
 	{
-		Line.Replace(TEXT("    "), TEXT("\t"));
+		Line = Line.Replace(TEXT("    "), TEXT("\t"));
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/CodeWriter.cpp:41]: Did this ever work?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 7]
```diff
@@ -305,17 +305,11 @@ bool FSpatialGDKEditor::LoadPotentialAssets(TArray<TStrongObjectPtr<UObject>>& O
...
 
-		const FString* GeneratedClassPathPtr = nullptr;
-
 		FAssetDataTagMapSharedView::FFindTagResult GeneratedClassFindTagResult = Data.TagsAndValues.FindTag("GeneratedClass");
 		if (GeneratedClassFindTagResult.IsSet())
 		{
-			GeneratedClassPathPtr = &GeneratedClassFindTagResult.GetValue();
-		}
-
-		if (GeneratedClassPathPtr != nullptr)
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp:316]: Do we no longer care about this failure case?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/CodeWriter.cpp (hunk 6 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/CodeWriter.cpp:41]: no, probably not

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp (hunk 7 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp:316]: I think this failure case was fully covered by `if (GeneratedClassFindTagResult.IsSet())`, since GeneratedClassPathPtr was being set to the address of a value we got for sure if the tag result was set.

**Review by simonsarginson — APPROVED** [review]
